### PR TITLE
Add support for marker filtering

### PR DESCRIPTION
**Pull request by ankusht-work**
-Add a context menu in the toolbar for filtering marker categories
-Add submenus dynamically based on the output selected.
-Add support for toggling.
-If more than one output is selected then the list will have a subset of all the categories
-Upon closing a output, the list will update accordingly
-The filter icon will disappear if the list is empty.
 
Signed-off-by: Ankush Tyagi <ankush.tyagi@ericsson.com>

### CODE REVIEW

File: packages/react-components/src/components/timegraph-output-component.tsx [hunk 1]
```diff
@@ -56,15 +58,18 @@ export class TimegraphOutputComponent extends AbstractTreeOutputComponent<Timegr
...
 
     constructor(props: TimegraphOutputProps) {
         super(props);
         this.state = {
             outputStatus: ResponseStatus.RUNNING,
             timegraphTree: [],
             collapsedNodes: [],
-            columns: []
+            columns: [],
+            annotationMarkers:[]
```
Comment (PatrickTasse) [inline packages/react-components/src/components/timegraph-output-component.tsx:72]: After a restart with trace and view already opened, there is no marker filter changed signal received, so the list of requested markers is empty, even though the filter shows all selected.

(this is a problem only after fixing not omitting requested_marker_categories parameter when the list is empty)

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-commands.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { Command, MenuPath } from '@theia/core';
+
+export const TRACE_VIEWER_TOOLBAR_COMMAND_FILTER: Command = {
+    id: 'trace.viewer.toolbar.filter',
+    label: 'Trace Viewer Toolbar Filter',
+    iconClass: 'fa fa-filter',
```
Comment (PatrickTasse) [inline theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-commands.ts:6]: Just checking, is this correct? I'm not familiar with this 'fa fa' thing...

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    registerToolbarItems(registry: TabBarToolbarRegistry): void {
+        registry.registerItem({
+            id: TRACE_VIEWER_TOOLBAR_COMMAND_FILTER.id,
+            isVisible: w => w instanceof TraceViewerWidget && this.outputAnnotationsMap.size>0,
+            // render() here is not a react component and hence need to disable the react display-name rule
+            // eslint-disable-next-line react/display-name
+            render: () => <div className="item enabled">
+                <div id="trace.viewer.toolbar.filter" className="fa fa-filter" title="Trace Viewer Toolbar filter" onClick={async (event: React.MouseEvent) => {
```
Comment (PatrickTasse) [inline theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx:60]: The title 'Trace Viewer Toolbar filter' is not specific enough for the user. It should be something like 'Markers filter'

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    initializeLayout(): void {
+        this.openView({ activate: false });
+    }
+
+    private doHandleAnnotationsFetchedSignal(payload: { [key: string]: { categoryName: string, toggleInd: boolean }[] }) {
+        Object.entries(payload).forEach(([key, value]) => {
+            if (value.length === 0) {
+                this.outputAnnotationsMap.delete(key);
```
Comment (PatrickTasse) [inline theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx:45]: Looks like there's only one global list. So when you have two traces opened with the same view output opened, and you remove the output from one trace, the global list is empty and the filter is disabled.

Then when you switch back to the other trace, the filter is still disabled even though that trace still has an output opened.

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

File: packages/react-components/src/components/data-providers/tsp-data-provider.ts [hunk 5]
```diff
@@ -47,11 +48,13 @@ export class TspDataProvider {
...
         this.totalRange = totalTimeRange.getEnd() - totalTimeRange.getstart();
-        let fetchParameters = QueryHelper.selectionTimeQuery(QueryHelper.splitRangeIntoEqualParts(1332170682440133097, 1332170682540133097, 1120), ids);
+        let fetchParameters = QueryHelper.selectionTimeQuery(QueryHelper.splitRangeIntoEqualParts(1332170682440133097, 1332170682540133097, 1120),
+                              ids, annotationMarkers && annotationMarkers.length>0 ? { 'requested_marker_categories': annotationMarkers }:{});
         if (viewRange && resolution) {
             const start = totalTimeRange.getstart() + viewRange.start;
             const end = totalTimeRange.getstart() + viewRange.end;
-            fetchParameters = QueryHelper.selectionTimeQuery(QueryHelper.splitRangeIntoEqualParts(Math.trunc(start), Math.trunc(end), resolution), ids);
+            fetchParameters = QueryHelper.selectionTimeQuery(QueryHelper.splitRangeIntoEqualParts(Math.trunc(start), Math.trunc(end), resolution),
+                              ids, annotationMarkers && annotationMarkers.length>0 ?{ 'requested_marker_categories': annotationMarkers }:{});
```
[the remark this answers is not in the document]
Comment (ankusht-work) [reply packages/react-components/src/components/data-providers/tsp-data-provider.ts]: Initialized requested_categories list to undefined. If it's undefined then remove the "requested_category" from the fetch parameter and the back-end api will return everything.

File: packages/react-components/src/components/data-providers/tsp-data-provider.ts [hunk 6]
```diff
@@ -33,7 +33,8 @@ export class TspDataProvider {
     }
 
     async getData(ids: number[], entries: TimeGraphEntry[], totalTimeRange: TimeRange,
-        viewRange?: TimelineChart.TimeGraphRange, resolution?: number): Promise<TimelineChart.TimeGraphModel> {
+
```
[the remark this answers is not in the document]
Comment (ankusht-work) [reply packages/react-components/src/components/data-providers/tsp-data-provider.ts]: fixed.

File: packages/react-components/src/components/timegraph-output-component.tsx (hunk 1 above)
Comment (ankusht-work) [reply packages/react-components/src/components/timegraph-output-component.tsx]: I am not seeing this issue. @bhufmann could you please confirm??

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-commands.ts (hunk 2 above)
Comment (ankusht-work) [reply theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-commands.ts:6]: Yes this is correct. 'fa fa' is font awesome thing :)

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx (hunk 3 above)
Comment (ankusht-work) [reply theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx]: Fixed

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx (hunk 4 above)
Comment (ankusht-work) [reply theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-toolbar-contribution.tsx:45]: Yeah that is a known issue. If you have multiple tabs open then the logic to add/remove filter categories based on the output id fails.

File: packages/react-components/src/components/timegraph-output-component.tsx (hunk 1 above)
Comment (bhufmann) [reply packages/react-components/src/components/timegraph-output-component.tsx]: @ankusht-work Not sure if it's the same issue, but I notice some inconsistency between the checked state in the menu and what is shown: 
1) Open a trace trace-1 and view with a marker category category-A
2) Deselect cat-A, view hides category-A
3) Open a second trace trace-2 and view with marker category category-A.  Menu shows category-A checked but view doesn't show markers of category-A

**Review by PatrickTasse — APPROVED** [review]
Issue to be created for handling categories state of multiple trace tabs.

File: packages/react-components/src/components/timegraph-output-component.tsx (hunk 1 above)
Comment (bhufmann) [reply packages/react-components/src/components/timegraph-output-component.tsx]: This has been fixed in the latest uploaded version [5cf9d08](https://github.com/theia-ide/theia-trace-extension/commit/5cf9d08716013d5b879c7b1c787791f42859f31c).

**Review by bhufmann — APPROVED** [review]
Looks good to me. Nice improvement to be able to hide markers. Thanks!

**Pull request merged**